### PR TITLE
Components: FormTokenField - add prop to allow saving of tokens onBlur

### PR DESCRIPTION
**Pull request by glendaviesnz**
## What?
Adds a `tokenizeOnBlur` prop.

## Why?
To allow consuming components to specify that any`incompleteTokenValue` is saved when the field loses focus. The main purpose for this is to prevent users saving the new [add patterns model](https://github.com/WordPress/gutenberg/pull/53835) with unsaved categories added.

## How?
If `tokenizeOnBlur` is `true` then `addNewToken` is run in `onBlur`

## Testing Instructions

- In the post editor add a new block and then in the block overflow menu choose the `Create pattern` menu option
- In the category input box add a category, and without hitting enter or `'` click outside the box so it loses focus
- Check that the category has been added and appears as token

## Screenshots or screencast 

Before:

https://github.com/WordPress/gutenberg/assets/3629020/d863e518-ac56-4fe9-a87c-0979aeaf57ee

After:

https://github.com/WordPress/gutenberg/assets/3629020/6ac41a4e-4867-4f4d-b4db-c7b1545c069d

### CODE REVIEW

File: packages/components/src/form-token-field/index.tsx
Comment (glendaviesnz) [inline packages/components/src/form-token-field/index.tsx]: the `isOnBlur` param is needed as the `setIsActive( false );` in the `onBlur` function is not registered until the next render so `isActive` is still `true` when `addNewToken` is called. There may be a better way of achieving this but couldn't think of one.

Comment (ciampo) [reply packages/components/src/form-token-field/index.tsx]: We could use `useEffect` and react to `isActive` becoming `false`, but the way this component is written is quite entangled, and therefore we'd still face issues.

I wonder if we could use the `hasFocus()` function and the `__experimentalAddOnBlur` prop in the `addNewToken` function instead of passing a `isOnBlur` argument around?

File: packages/components/src/form-token-field/types.ts
Comment (ciampo) [inline packages/components/src/form-token-field/types.ts]: Since it's been established that `__experimental` prop prefixes don't really make a prop experimental after it's merged in WP core, I wonder if we should just remove that prefix.

Also, since there is already a `tokenizeOnSpace` prop, should we call this prop `tokenizeOnBlur` ?

**Review by ciampo — COMMENTED** [review]
Left a couple of comments, but overall this PR makes sense to me.

It would be great if you could also add unit tests to cover the new behavior

File: packages/components/src/form-token-field/types.ts
Comment (glendaviesnz) [reply packages/components/src/form-token-field/types.ts]: good point, have switched to `tokenizeOnBlur`

File: packages/components/src/form-token-field/index.tsx
Comment (glendaviesnz) [reply packages/components/src/form-token-field/index.tsx]: [Refactored this here](https://github.com/WordPress/gutenberg/pull/53976/commits/bf20870aec7677b9f194baf4f1fecd14f2c6ac1a) to get rid of `isOnBlur` as suggested and it seems to work, but will do some more testing tomorrow to make sure I didn't miss any edge cases.

**Comment by glendaviesnz** [comment]
> It would be great if you could also add unit tests to cover the new behavior

If you are happy with the general approach with the latest changes I will go ahead and add some additional unit tests tomorrow  @ciampo

**Comment by ciampo** [comment]
Yup, the overall approach looks good to me!

**Comment by glendaviesnz** [comment]
Unit tests have been added.

File: packages/components/src/form-token-field/README.md
Comment (aaronrobertshaw) [inline packages/components/src/form-token-field/README.md]:
```suggestion
-   `tokenizeOnBlur` - If true, add any incompleteTokenValue as a new token when the field loses focus.
```

File: packages/components/src/form-token-field/test/index.tsx
Comment (aaronrobertshaw) [inline packages/components/src/form-token-field/test/index.tsx]: We can avoid needing to use `act` and bring the new test in line with others in this file, see comment below.

File: packages/components/src/form-token-field/test/index.tsx
Comment (aaronrobertshaw) [inline packages/components/src/form-token-field/test/index.tsx]:
```suggestion
			await user.click( document.body );
```

I believe the current preference is to test as much as possible from the user's perspective. The suggested change above is how existing tests around blur in this file approach this. Making the approach consistent is probably a good thing too.

File: packages/components/src/form-token-field/test/index.tsx
Comment (aaronrobertshaw) [inline packages/components/src/form-token-field/test/index.tsx]:
```suggestion
			await user.click( document.body );
```

Same as above.

File: packages/components/src/form-token-field/test/index.tsx
Comment (aaronrobertshaw) [inline packages/components/src/form-token-field/test/index.tsx]:
```suggestion
		it( "should not add a token with the input's value when tokenizeOnBlur is not set and pressing the tab key", async () => {
```

We should probably make this test description more accurate as well.

File: packages/components/src/form-token-field/types.ts
Comment (aaronrobertshaw) [inline packages/components/src/form-token-field/types.ts]:
```suggestion
	 * If true, add any incompleteTokenValue as a new token when the field loses focus.
```

Small typo tweak and making this consistent with the readme.

**Review by aaronrobertshaw — APPROVED** [review]
Thanks for this one @glendaviesnz, I often trip over the incomplete values not being tokenized, especially when testing pattern categories and taxonomies lately.

✅ Storybook examples function as intended
✅ The token field for pattern categories works as advertised
✅ Unit tests pass

I've left some minor tweaks and typo nits via inline comments. Also, the PR description is out of date after the renaming of the new prop, we should bring that up-to-date as well. 

With the above tweaks made, I think this will be ready to land. Nice work!

<details>
<summary>Here's a diff with suggested tweaks and an extra typo fix that I couldn't add to this review in case it helps.</summary>

```diff
diff --git a/packages/components/src/form-token-field/README.md b/packages/components/src/form-token-field/README.md
index b442e42c31..9e31dcafeb 100644
--- a/packages/components/src/form-token-field/README.md
+++ b/packages/components/src/form-token-field/README.md
@@ -62,7 +62,7 @@ The `value` property is handled in a manner similar to controlled form component
 -   `__experimentalValidateInput` - If passed, all introduced values will be validated before being added as tokens.
 -   `__experimentalAutoSelectFirstMatch` - If true, the select the first matching suggestion when the user presses the Enter key (or space when tokenizeOnSpace is true).
 -   `__nextHasNoMarginBottom` - Start opting into the new margin-free styles that will become the default in a future version, currently scheduled to be WordPress 6.5. (The prop can be safely removed once this happens.)
--   `tokenizeOnBlur` - If true adds any incompleteTokenValue as a new token when field loses focus.
+-   `tokenizeOnBlur` - If true, adds any incompleteTokenValue as a new token when field loses focus.
 
 ## Usage
 
diff --git a/packages/components/src/form-token-field/test/index.tsx b/packages/components/src/form-token-field/test/index.tsx
index 81e61ea186..c6a8397cb2 100644
--- a/packages/components/src/form-token-field/test/index.tsx
+++ b/packages/components/src/form-token-field/test/index.tsx
@@ -9,7 +9,6 @@ import {
 	within,
 	getDefaultNormalizer,
 	waitFor,
-	act,
 } from '@testing-library/react';
 import userEvent from '@testing-library/user-event';
 import type { ComponentProps } from 'react';
@@ -219,9 +218,8 @@ describe( 'FormTokenField', () => {
 
 			// Add 'grapefruit' token by typing it and check blur of field does not tokenize it.
 			await user.type( input, 'grapefruit' );
-			act( () => {
-				input.blur();
-			} );
+			await user.click( document.body );
+
 			expect( onChangeSpy ).toHaveBeenCalledTimes( 0 );
 			expectTokensNotToBeInTheDocument( [ 'grapefruit' ] );
 
@@ -235,17 +233,15 @@ describe( 'FormTokenField', () => {
 
 			// Add 'grapefruit' token by typing it and check blur of field tokenizes it.
 			await user.type( input, 'grapefruit' );
+			await user.click( document.body );
 
-			act( () => {
-				input.blur();
-			} );
 			expect( onChangeSpy ).toHaveBeenNthCalledWith( 1, [
 				'grapefruit',
 			] );
 			expectTokensToBeInTheDocument( [ 'grapefruit' ] );
 		} );
 
-		it( "should not add a token with the input's value when pressing the tab key", async () => {
+		it( "should not add a token with the input's value when tokenizeOnBlur is not set and pressing the tab key", async () => {
 			const user = userEvent.setup();
 
 			const onChangeSpy = jest.fn();
@@ -254,7 +250,6 @@ describe( 'FormTokenField', () => {
 
 			const input = screen.getByRole( 'combobox' );
 
-			// Add 'orange' token by typing it and pressing enter to tokenize it.
 			await user.type( input, 'grapefruit' );
 			await user.tab();
 			expect( onChangeSpy ).toHaveBeenCalledTimes( 0 );
diff --git a/packages/components/src/form-token-field/types.ts b/packages/components/src/form-token-field/types.ts
index be34eadd1c..92e2d51fe1 100644
--- a/packages/components/src/form-token-field/types.ts
+++ b/packages/components/src/form-token-field/types.ts
@@ -166,7 +166,7 @@ export interface FormTokenFieldProps
 	 */
 	__next40pxDefaultSize?: boolean;
 	/**
-	 * If true, the select the first matching suggestion when the user presses
+	 * If true, then select the first matching suggestion when the user presses
 	 * the Enter key (or space when tokenizeOnSpace is true).
 	 *
 	 * @default false
@@ -183,7 +183,7 @@ export interface FormTokenFieldProps
 	 */
 	__nextHasNoMarginBottom?: boolean;
 	/**
-	 * If true add any incompleteTokenValue as a new token.
+	 * If true, add any incompleteTokenValue as a new token when the field loses focus.
 	 *
 	 * @default false
 	 */

```

</details>

File: packages/components/src/form-token-field/test/index.tsx
Comment (glendaviesnz) [reply packages/components/src/form-token-field/test/index.tsx]: That test is actually the existing one following the one I added, and it does test the tab key I think 😃

Comment (aaronrobertshaw) [reply packages/components/src/form-token-field/test/index.tsx]: It is the existing test, however, it should only not add the token when hitting tab when tokenizeOnBlur is not set, hence the clarification in the description. If you add `tokenizeOnBlur` to that test's component render, it will fail.

Am I missing something?

File: packages/components/src/form-token-field/README.md
Comment (glendaviesnz) [reply packages/components/src/form-token-field/README.md]: Fixed

File: packages/components/src/form-token-field/test/index.tsx
Comment (glendaviesnz) [reply packages/components/src/form-token-field/test/index.tsx]: Doh! That is the obvious solution that I couldn't see for looking this morning, thanks, updated in both places.

File: packages/components/src/form-token-field/types.ts
Comment (glendaviesnz) [reply packages/components/src/form-token-field/types.ts]: fixed.

File: packages/components/src/form-token-field/test/index.tsx
Comment (glendaviesnz) [reply packages/components/src/form-token-field/test/index.tsx]: > Am I missing something?

No but I was, I see what you mean now, fixed.

**Review by aaronrobertshaw — APPROVED** [review]
Latest changes LGTM 🚀

**Comment by glendaviesnz** [comment]
Merging this into #53835 first as it was branched off there, and we will do another full text of this functionality there before merging into trunk. Igoring php test failures as unrelated and should be fixed after rebase with trunk

**Comment by ciampo** [comment]
I personally prefer merging PRs separately especially when they affect more than one package, since they are easier to find/understand/revert if needed. But I understand if that was particularly tricky this time.

Thank you anyway @glendaviesnz for the heads-up!

**Comment by glendaviesnz** [comment]
> I personally prefer merging PRs separately especially when they affect more than one package,

Yeh, after rethinking this, the better approach will be to wait until the rest of the pattern category feature work is merged into trunk, and then merge this to trunk as a separate PR, so I have reverted that merge and added 
https://github.com/WordPress/gutenberg/pull/54445 which I will rebase to trunk once all the other work is merged. Sorry about the confusion and thanks for the ping @ciampo

**Pull request merged**
